tmt/steps/provision/beaker_mrack.py

```diff
@@ -0,0 +1,532 @@
+        mrack.logger.removeHandler(mrack.console_handler)
+
+        self._guest = guest
+
+        # use global context class
+        global_context = mrack.context.global_context
+
+        #  FIXME 1.12.1 should work with conf only TODO: load from /etc/tmt/mrack.conf
+        mrack_config = "mrack.conf"
+        global_context.init(mrack_config)
+
+        self._mrack_transformer = TmtBeakerTransformer()
+        await self._mrack_transformer.init(global_context.PROV_CONFIG, {})
+        self._mrack_provider = self._mrack_transformer._provider
+        self._mrack_provider.poll_sleep = DEFAULT_PROVISION_TICK
+
+    @async_run
+    async def create(
+            self,
+            data: Dict[str, Any],
+            ) -> Any:
+        """
+        Create - or request creation of - a resource using mrack up.
+
+        :param data: optional key/value data to send with the request.
+
+        """
+        req = self._mrack_transformer.create_host_requirement(data)
+        self._bkr_job_id, self._req = await self._mrack_provider.create_server(req)
+        return self._mrack_provider._get_recipe_info(self._bkr_job_id)
+
+    @async_run
+    async def inspect(
+            self,
+            ) -> Any:
+        """
+        Inspect a resource.  # kinda wait till provisioned
+
+        """
+        return self._mrack_provider._get_recipe_info(self._bkr_job_id)
+
+    @async_run
+    async def delete(  # destroy
+            self,
+            ) -> Any:
+        """
+        Delete - or request removal of - a resource.
+
+        """
+        return await self._mrack_provider.delete_host(self._bkr_job_id, None)
+
+
+class GuestBeaker(tmt.GuestSsh):
+    """
+    Beaker guest instance
+
+    The following keys are expected in the 'data' dictionary:
+    """
+    _data_class = BeakerGuestData
+
+    # Guest request properties
+    arch: str
+    image: str
+    hardware: Dict[str, Any] = {}
+    pool: Optional[str]
+    keyname: str
+    user_data: Dict[str, str]
+
+    # Provided in Beaker response
+    guestname: Optional[str]
+
+    # Timeouts and deadlines
+    provision_timeout: int
+    provision_tick: int
+    _api: Optional[BeakerAPI] = None
+
+    @property
+    def api(self) -> BeakerAPI:
+        if self._api is None:
+            self._api = BeakerAPI(self)
+
+        return self._api
+
+    def _create(self) -> None:
+        environment: Dict[str, Any] = {
+            'hw': {
+                'arch': self.arch
+                },
+            'os': {
+                'compose': self.image
+                }
+            }
+
+        data: Dict[str, Any] = {
+            'environment': environment,
+            'keyname': self.keyname,
+            'user_data': self.user_data
+            }
+
+        if self.pool:
+            environment['pool'] = self.pool
+
+        if self.hardware is not None:
+            assert isinstance(self.hardware, dict)
+
+            environment['hw']['constraints'] = self.hardware
```

Comment:
The `environment` name seems to be unused - looks like a copy & paste from `artemis` plugin, and it's not needed for Beaker.

Reply:
Well, it is used. Sorry about that. But. `environment.hw.constraints` is/was constructed by Artemis plugin, because it's what Artemis API accepts - since your plugin does not deal with Artemis plugin, you can drop the extra layers, and just propagate `self.hardware` directly.

Reply:
rewritten

Reply:
I still see `hw.get("constraints", {})` under `_translate_tmt_hw`. We don't have any `constraints` key in the [hardware spec](https://tmt.readthedocs.io/en/latest/spec/hardware.html). Or have I missed anything?